clang/lib/Sema/SemaBounds.cpp

```diff
@@ -1238,6 +1376,175 @@ namespace {
+                                DeclaredUpperPos, ObservedUpperPos,
+                                EquivExprs, FreeVariables)) {
+          Cause = CombineFailures(Cause, ProofFailure::HasFreeVariables);
+          return ProofResult::False;
+        }
+
+        return ProofResult::Maybe;
+      }
+
+      // GetFreeVariables gathers "free variables" in SrcVars.
+      //
+      // Given two variable sets SrcVars and DstVars, and a set of equivalent
+      // sets of Expr EquivExprs. A variable V in SrcVars is *free* if the two
+      // conditions are met:
+      //   1. either (1) V is not an integer type, or (2) there is no set in
+      //   EquivExprs that contains both V and a constant (i.e. an integer-typed
+      //   value); and
+      //   2. for each variable U in DstVars, there is not set in EquivExprs
+      //   that contains both V and U.
+      //
+      // GetFreeVariables returns true if any free variable is found in SrcVars,
+      // and appends the free variables to FreeVariables.
+      static bool GetFreeVariables(Sema &S, const EqualExprTy &SrcVars,
+                                   const EqualExprTy &DstVars,
+                                   EquivExprSets *EquivExprs,
+                                   EqualExprTy &FreeVariables) {
+        bool HasFreeVariables = false;
+
+        // Gather free variables.
+        for (const auto &SrcV : SrcVars) {
+          if (IsEqualToConstant(S, SrcV, EquivExprs))
+            continue;
+          auto It = DstVars.begin();
+          for (; It != DstVars.end(); It++) {
+            if (EqualValue(S.Context, SrcV, *It, EquivExprs))
+              break;
+          }
+
+          // Find no constant or no match for SrcV in EquivExprs.
+          if (It == DstVars.end()) {
+            HasFreeVariables = true;
+            FreeVariables.push_back(SrcV);
+          }
+        }
+        return HasFreeVariables;
+      }
+
+      // Check free variables between E1 and E2. Append any found free variables
+      // to FreeVars with each free variable in E1 (resp. E2) having Pos1 (resp.
+      // Pos2).
+      static bool CheckFreeVarInExprs(Sema &S, Expr *E1, Expr *E2,
+                                      FreeVariablePosition Pos1,
+                                      FreeVariablePosition Pos2,
+                                      EquivExprSets *EquivExprs,
+                                      FreeVariableListTy &FreeVars) {
+        // If E1 or E2 accesses memory via poiter, we skip because we cannot
+        // determine aliases for two indirect accesses soundly yet.
+        if (ReadsMemoryViaPointer(E1) || ReadsMemoryViaPointer(E2))
+          return false;
+
+        bool HasFreeVariables = false;
+        EqualExprTy Vars1 = CollectVariableSet(S, E1);
+        EqualExprTy Vars2 = CollectVariableSet(S, E2);
+
+        // EquivVars holds sets of DeclRefExpr and IntegerLiteral filtered from
+        // EquivExprs.
+        EquivExprSets EquivVars;
```

Comment:
An implication of my comment is that we shouldn't be filtering EquivVars.  We could be removing expressions that have equality relationships and that use variables.